src/validation.cpp

```diff
@@ -4106,7 +4107,6 @@ void UnloadBlockIndex(CTxMemPool* mempool, ChainstateManager& chainman)
     LOCK(cs_main);
     chainman.Unload();
     if (mempool) mempool->clear();
-    g_versionbitscache.Clear();
```

Comment:
Note that there shouldn't be any logical need for this to be cleared here -- the cached information doesn't become invalid when we load up different blocks, only if the consensus parameters are changed.

Reply:
Is this the same case for `warningcache` as well? If so, perhaps we'll just move the clearing logic to `SelectParams` and not move these two structures into `BlockManager`?

Reply:
Yes, `warningcache` only relies on the chain history, consensus params, and versionbitscache state. Moving the clearing logic to `SelectParams` would introduce a circular dependency though, I think?

Reply:
Here's a rebase of this that replaces moving g_versionbits with just removing `g_versionbitscache.Clear()` from `UnloadBlockIndex` -- https://github.com/ajtowns/bitcoin/commits/202201-vb-unloadblockindex ... What do you think about doing that instead? I've got some updates for the versionbits code I'd like to PR, and would prefer to minimise conflicts with this one...

Reply:
If `warningcache` and `g_versionbitscache` are both only dependent on consensus parameters, then perhaps (perhaps not in this PR) we should move them to `ChainstateManager` or `NodeContext`? At the very least we would need to clear them in `~ChainTestingSetup` so that tests with different consensus params won't get confused.

Reply:
Moving them somewhere rather than having them be globals would be good eventually; I had a look at moving them into `ChainstateManager` but it felt a bit too invasive when what I was hoping to do was avoid/defer introducing conflicts...

The patch above clears `g_versionbitscache` in the `BasicTestingSetup` constructor instead (where `SelectParams` is invoked), which I think should be good enough? It's also already cleared explicitly in `versionbits_tests.cpp` which manually changes the consensus params being used.

Reply:
#24595 moves `g_versionbitscache` into `ChainstateManager` and builds on this PR, so hopefully resolves this thread